From 607ba827f2d4d51ba38acc0ed263dc3d6eb1d14e Mon Sep 17 00:00:00 2001
From: Sarika Sharma <quic_sarishar@quicinc.com>
Date: Thu, 11 Jul 2024 15:14:01 +0530
Subject: [PATCH] wifi: mac80211: add warn_on for sending unexpected 4addr null
 frame using eapol M2

Currently, WDS (4addr) AP creates AP/VLAN interface upon receiving
4addr Null data frame from WDS (4addr) STA immediately after
association response success.

If at any chance above didn't succeed, WDS (4addr) AP tries to
create AP/VLAN interface upon receiving any first 4addr data
frame(M2 EAPOL). This is achieved by notifying hostapd
with NL80211_CMD_UNEXPECTED_4ADDR_FRAME for AP/VLAN interface
creation.

Hence, add WARN_ON_ONCE if AP_VLAN creation happened from eapol M2
packet.

Patch-dependency: 904-wifi-mac80211-fix-EAPOL-M2-drop-in-WDS-4addr-AP.patch
		  904-wifi-ath12k-Add-support-to-process-WDS-null-frame.patch

Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 net/mac80211/rx.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -3358,10 +3358,16 @@ ieee80211_rx_h_data(struct ieee80211_rx_
 	if (ieee80211_has_a4(hdr->frame_control) &&
 	    sdata->vif.type == NL80211_IFTYPE_AP) {
 		if (rx->sta &&
-		    !test_and_set_sta_flag(rx->sta, WLAN_STA_4ADDR_EVENT))
+		    !test_and_set_sta_flag(rx->sta, WLAN_STA_4ADDR_EVENT)) {
 			cfg80211_rx_unexpected_4addr_frame(
 				rx->sdata->dev, rx->sta->sta.addr, GFP_ATOMIC,
 				rx->link_id);
+			if (sdata->vif.offload_flags & IEEE80211_OFFLOAD_ENCAP_4ADDR) {
+				pr_warn("4addr non-null data frame: %d with frame_control: %x",
+					port_control, hdr->frame_control);
+				WARN_ON_ONCE(1);
+			}
+		}
 		if (!port_control)
 			return RX_DROP_MONITOR;
 	}
